### PR TITLE
Money's internal amount should be a string

### LINKED ISSUE #133: Careful with string vs int
Now that there are no more integer limitations (#115), we have to be careful with `int` and `string` usage vs strict comparison. This is something that I did not cover in my PR but now realize that should be taken care of. Especially with these lines.

- [isZero](https://github.com/mathiasverraes/money/blob/nextrelease/src/Money.php#L421)
- [isPositive](https://github.com/mathiasverraes/money/blob/nextrelease/src/Money.php#L431)
- [isNegative](https://github.com/mathiasverraes/money/blob/nextrelease/src/Money.php#L441)
- [equals](https://github.com/mathiasverraes/money/blob/nextrelease/src/Money.php#L133)

These lines are now buggy, because the following will return false.

```php
(new Money(1, new Currency('EUR')))->equals(new Money('1', new Currency('EUR')))
```

I guess we should convert the amount to a string in the constructor with `(string)`. PR is on the way.

### PR DESCRIPTION
**Pull request by frederikbosch**
Fixes #133.

### CODE REVIEW

**Comment by frederikbosch** [comment]
In order to create a `MoneyParser` PR, I need this to be merged. I just improved more docblock and squashed my commit.

File: src/Calculator/PhpCalculator.php [hunk 1]
```diff
@@ -124,6 +124,19 @@ private function assertIntegerBounds($amount)
...
 
     /**
+     * Casts an amount to string ensuring that the internal
+     * amount value of Money is a string
+     *
+     * @param int $amount
+     *
+     * @return int
+     */
+    private function castString($amount)
```
Comment (sagikazarmark) [inline src/Calculator/PhpCalculator.php]: Do we really need a separate method for that?

Comment (frederikbosch) [inline src/Calculator/PhpCalculator.php]: No, we really do not need it. Removed it, and squashed again.

File: src/Money.php [hunk 2]
```diff
@@ -47,7 +47,7 @@
     ];
 
     /**
-     * @param int      $amount   Amount, expressed in the smallest units of $currency (eg cents)
+     * @param int|string $amount   Amount,expressed in the smallest units of $currency (eg cents)
```
Comment (sagikazarmark) [inline src/Money.php]: I would say we need that space after the comma.

**Comment by sagikazarmark** [comment]
Well, I tend accept this merge, but still need to get used to the fact that amount will be stored as string.

**Comment by frederikbosch** [comment]
@sagikazarmark I agree. It is a big change. It was more comprehensive than I thought at first.

**Comment by sagikazarmark** [comment]
Merging the deprecation PR caused some conflicts.

**Comment by frederikbosch** [comment]
Yeah, and now I need to rebase? :) Still don't know how, sorry about that...

**Comment by sagikazarmark** [comment]
No problem, I will do it.

**Comment by sagikazarmark** [comment]
Conflicts solved in #136

**Comment by josecelano** [comment]
Some months ago I was testing how to deal with bit integers and I did this fork:

https://github.com/josecelano/money/blob/master/lib/Money/BigMoney.php

Instead of using integer or string for the amount I used another class. In this implementation Money has not the responsibility to deal with bit integers and it delegates to another class. The only problem is that library uses only `bcmath` extension. But maybe there is another similar class which uses `calculators`as you did it.

**Comment by frederikbosch** [comment]
@josecelano `bcmath` is the only option for `PreciseMoney`. I rather use that term than `BigMoney`, because we already have `BigMoney` because there is no more integer limit. `GMP` only uses integers, not decimals, see the [RFC for floating points integration](https://wiki.php.net/rfc/gmp-floating-point) in PHP.

The solution I will propose soon will have the following structure.
- `PreciseCalculator` with methods like `roundPrecisely`
- `PreciseMoney`
- `BcMathCalculator` implementing `PreciseCalculator`
- I still have to think have to deal with formatting and parsing.

When the RFC for `GMP` lands, we only have to implement `PreciseCalculator` for GMP.

**Comment by josecelano** [comment]
Sorry,there was definitely a misunderstanding. What I called BigMoney is an implementation of Money which allows big integers, not decimals. I used Decimal class but internally I force the constrain of not using decimals, I mean all values would have no deciaml part even if that class allows it. But the idea I want to expose is you could use another class to store the amount and delegate to that class the responsabiliti of dealing with bit integers. In fact, I was looking for a `BigInteger` class but I do nto remember why finally I decided to use `Decimal` for that test.

**Comment by sagikazarmark** [comment]
Yeah, BigMoney casued a lot of confusion in me as well. But it refers to the precision of the value, not the integer size.

**Pull request closed without merging**